File: bindings/go/execute.go
package memless

import "strings"

// Execute runs a write sql against the instance and returns the affected row
// count. It returns a RefusalError on a validation or disk failure, and a
// FaultError on a boundary or internal fault.
func (i *Instance) Execute(sql string) (uint64, error) {
	if strings.IndexByte(sql, 0) >= 0 {
		return 0, &FaultError{Status: statusInvalidArgument, Message: "sql contains a NUL byte"}
	}
	if err := ensureLoaded(); err != nil {
		return 0, err
	}
	var affected uint64
	var message *byte
	status := memlessExecute(i.handle, sql, &affected, &message)
	return interpretExecute(status, affected, message)
}

func interpretExecute(status int32, affected uint64, message *byte) (uint64, error) {
	text := takeMessage(message)
	if status == statusOk {
		return affected, nil
	}
	if status == statusRefused {
		return 0, &RefusalError{Message: text}
	}
	return 0, &FaultError{Status: status, Message: text}
}

File: bindings/php/src/Execute.php
<?php

declare(strict_types=1);

namespace Memless;

/**
 * Runs one SQL write or transaction verb against an instance handle: calls
 * memless_execute, then returns the affected row count on Ok (0 for a verb),
 * throws the refusal message on Refused (a validation, disk or transaction-guard
 * failure), a fault otherwise.
 */
final class Execute
{
    private const STATUS_OK = 0;

    private const STATUS_REFUSED = 1;

    public static function run(int $handle, string $sql): int
    {
        if (strpos($sql, "\0") !== false) {
            throw new \InvalidArgumentException('sql contains a NUL byte');
        }

        $ffi = Library::ffi();
        [$status, $affected, $message] = self::call($ffi, $handle, $sql);

        return self::outcome($status, $affected, $message);
    }

    private static function call(\FFI $ffi, int $handle, string $sql): array
    {
        $outAffected = $ffi->new('uint64_t');
        $outMessage = $ffi->new('char*');
        $status = $ffi->memless_execute($handle, $sql, \FFI::addr($outAffected), \FFI::addr($outMessage));

        return [$status, (int) $outAffected->cdata, Message::take($ffi, $outMessage)];
    }

    private static function outcome(int $status, int $affected, string $message): int
    {
        if ($status === self::STATUS_OK) {
            return $affected;
        }

        if ($status === self::STATUS_REFUSED) {
            throw new MemlessRefusal($message);
        }

        throw new \LogicException("memless execute fault ({$status}): {$message}");
    }
}
